Encrypt Node_id and Tor_id at rest

SWvheerden opened the pull request:
## Description
This PR adds in the NodeID and TorID to the wallet's database as encrypted fields 

## Motivation and Context
This allows us to protect the sensitive private keys for the Tari Comms and Tor comms as they are not stored in a plain text file at rest anymore. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [ ] I'm merging against the `development` branch.
* [ ] I ran `cargo-fmt --all` before pushing.
* [ ] I ran `cargo test` successfully before submitting my PR.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

stringhandler commented on applications/tari_console_wallet/src/init/mod.rs:
```diff
@@ -206,21 +222,83 @@ pub async fn init_wallet(
+            },
+        }
+    };
+    debug!(target: LOG_TARGET, "after testnet case");
```
Debugging code?

stringhandler commented on applications/tari_console_wallet/src/init/mod.rs:
```diff
@@ -206,21 +222,83 @@ pub async fn init_wallet(
         "Databases Initialized. Wallet encrypted? {}.", wallet_encrypted
-    );
-
+    ); // TODO remove after next TestNet
```
Perhaps some more detail here?

stringhandler commented on applications/tari_console_wallet/src/init/mod.rs:
```diff
@@ -206,21 +222,83 @@ pub async fn init_wallet(
+        }
+    };
+    debug!(target: LOG_TARGET, "after testnet case");
+    let _ = wallet_backend.fetch(&DbKey::Identity);
```
Duplicate code

stringhandler commented on applications/tari_console_wallet/src/init/mod.rs:
```diff
@@ -206,21 +222,83 @@ pub async fn init_wallet(
-    );
-
+    ); // TODO remove after next TestNet
+    if node_identity.is_some() {
```
clippy usually picks this up as `if let Some(x) = ... {`

SWvheerden replied:
Did all clippy comments

stringhandler commented on base_layer/wallet/src/storage/database.rs:
```diff
@@ -47,12 +52,20 @@ pub trait WalletBackend: Send + Sync + Clone {
+    CommsAddress(Multiaddr),
+    CommsFeatures(u64),
+    Identity(NodeIdentity),
+    TorID(TorIdentity),
```
```suggestion
    TorId(TorIdentity),
```

stringhandler commented on base_layer/wallet/src/storage/sqlite_db.rs:
```diff
@@ -204,31 +230,93 @@ impl WalletSqliteDatabase {
+
+    fn get_tor_id(&self, conn: &SqliteConnection) -> Result<Option<DbValue>, WalletStorageError> {
+        let cipher = acquire_read_lock!(self.cipher);
+        if let Some(key_str) = WalletSettingSql::get(format!("{}", DbKey::TorID), &conn)? {
```
nit: `.to_string()`

SWvheerden replied:
Removed all format!("{}" and replaced with to_string()

stringhandler commented on base_layer/wallet/src/storage/sqlite_db.rs:
```diff
@@ -344,6 +485,21 @@ impl WalletBackend for WalletSqliteDatabase {
+        match key_str {
+            Some(v) => {
+                let decrypted_key_bytes = decrypt_bytes_integral_nonce(&cipher, from_hex(v.as_str())?)
+                    .map_err(|_| WalletStorageError::AeadError("Decryption Error".to_string()))?;
```
Swallowing an error here

SWvheerden replied:
Changed the error type here to not swallow any errors anymore

stringhandler reviewed (commented):
Happy in general, but a couple of clippy warnings and style nits. There are also a number of places where you have swallowed errors

philipr-za commented on applications/tari_console_wallet/src/init/mod.rs:
```diff
@@ -206,21 +219,84 @@ pub async fn init_wallet(
+    // If we know node_identity is not passed in anymore we dont have to check if we need to write it to db. This
+    // assumes that if its passed in we need to still save it the database.
+    if node_identity.is_some() {
+        match Arc::try_unwrap(node_identity.unwrap()) {
```
Pretty sure you can do away with all of this unwrapping stuff and just go:
```Rust
(*Arc::make_mut(&mut node_identity)).clone()
```
Assuming you use the suggested `if let Some(x)` above to get inside the option

SWvheerden replied:
Clippy was happy with this one,  but I updated the code with (*Arc::make_mut(&mut node_identity)).clone()

philipr-za commented on base_layer/wallet/src/storage/sqlite_db.rs:
```diff
@@ -204,33 +229,99 @@ impl WalletSqliteDatabase {
+                    let tor_json = std::str::from_utf8(&decrypted_key_bytes)
+                        .map_err(|_| WalletStorageError::AeadError("Decryption Error".to_string()))?
+                        .to_string();
+                    // Encryption and decryption breaks the json string, so we fix it.
```
This feels weird, the encryption should return the bytes stream byte for byte?

SWvheerden replied:
Its because we are encoding a string
so this:
"{\"private_key\":{\"Ed25519V3\":\"aFn0Bod+Vvq8/obUEbVPVsa9Vdk9p44FOavDKzJVoU72dyJ65k9Zy6+NuWrwI1uI51ApzG5IjDUH2NjW9fKmoA==\"},\"service_id\":\"pqdggzs7wlxguwdyynojpez3a7tfzo2zacks2ppivyp4pnsu7lmkpzqd\",\"onion_port\":18101}"
Becomes:
"\"{\\"private_key\\":{\\"Ed25519V3\\":\\"aFn0Bod+Vvq8/obUEbVPVsa9Vdk9p44FOavDKzJVoU72dyJ65k9Zy6+NuWrwI1uI51ApzG5IjDUH2NjW9fKmoA==\\"},\\"service_id\\":\\"pqdggzs7wlxguwdyynojpez3a7tfzo2zacks2ppivyp4pnsu7lmkpzqd\\",\\"onion_port\\":18101}\""

stringhandler replied:
Why not serialize it as bytes with serde?

SWvheerden replied:
I will try and do that when saving as encrypted

stringhandler commented on base_layer/wallet/src/storage/sqlite_db.rs:
```diff
@@ -204,33 +229,99 @@ impl WalletSqliteDatabase {
+                None => CommsSecretKey::from_hex(key_str.as_str())?,
+                Some(cipher) => {
+                    let decrypted_key_bytes = decrypt_bytes_integral_nonce(&cipher, from_hex(key_str.as_str())?)
+                        .map_err(|_| WalletStorageError::AeadError("Decryption Error".to_string()))?;
```
Still swallowing this error

SWvheerden replied:
I removed these all as well.

stringhandler commented on base_layer/wallet/src/storage/sqlite_db.rs:
```diff
@@ -243,6 +334,27 @@ impl WalletBackend for WalletSqliteDatabase {
+                    Some(DbValue::CommsSecretKey(v)) => v,
+                    _ => return Ok(None),
+                };
+                let address = match self.get_comms_address(&conn)? {
```
Why not rather return a valid `Option<CommsAddress>` here instead of wrapping it in a `DbValue`? Seems to add a lot of code in the checking stage

philipr-za replied:
This is part of this modular storage backend pattern that we conceived in the very beginning of the project to allow for multiple backends to be written that implement the `fetch(...)` and `write(...)` methods of the trait. But I agree I think the boilerplate this creates is not worth the benefit for a potential easier future integration of some other backend. Though removing it now is a cleanup refactor in itself so 🤷‍♂️

stringhandler replied:
Fair, but, I was more referring to the fact that `get_comms_address` could be written to return the result in the correct type (i.e. `CommsAddress`), and only the instance that needed to return a `DbValue` (i.e. `fetch`) wrap it.

SWvheerden replied:
This is the same schematics we use all over the code base, from the BaseNode to the wallet for db write transactions.  I only did it this way to be consistent with the rest of the code.

SWvheerden replied:
`fn get_comms_address` is a private function. 
You either have to unwrap the option here, and use above straight
OR
Use here straight, and wrap above

stringhandler replied:
I disagree that it is like that in the code base, and lines 330 and 324 are both examples of doing it correctly, and `get_comms_address` is a new method created in this PR. I would say change line 337 to 
```
DbKey::CommsAddress => self.get_comms_address(&conn)?.map(|v| DbValue::CommsAddress(v)),
```
and `get_comms_address` to:
```
fn get_comms_address(&self, conn: &SqliteConnection) -> Result<Option<CommsAddress>, WalletStorageError> 
```

stringhandler commented on base_layer/wallet/src/storage/sqlite_db.rs:
```diff
@@ -187,6 +192,26 @@ impl WalletSqliteDatabase {
+        let cipher = acquire_read_lock!(self.cipher);
+        WalletSettingSql::new(DbKey::CommsPublicKey.to_string(), node_id.public_key().to_hex()).set(&conn)?;
+        WalletSettingSql::new(DbKey::CommsAddress.to_string(), node_id.public_address().to_string()).set(&conn)?;
+        WalletSettingSql::new(DbKey::CommsFeatures.to_string(), node_id.features().bits().to_string()).set(&conn)?;
```
Why are features saved to the DB? Are these configurable?

SWvheerden replied:
features is, 
The other two is purely convenance so that we dont have to recalc them. I added everything from the node_id.json file.

stringhandler commented on base_layer/wallet_ffi/src/lib.rs:
```diff
@@ -2339,7 +2340,7 @@ pub unsafe extern "C" fn transport_tcp_create(
     control_server_address: *const c_char,
     tor_cookie: *const ByteVector,
-    tor_identity: *const ByteVector,
+    _tor_identity: *const ByteVector,
```
Probably best to just remove it

philipr-za replied:
careful, this will break the Swift and JNI wrappers if we don't update those at the same time.

philipr-za commented on base_layer/wallet_ffi/src/lib.rs:
```diff
@@ -2951,6 +2970,15 @@ pub unsafe extern "C" fn wallet_create(
+                        if let Err(e) = runtime.block_on(w.db.set_tor_identity(hs.tor_identity().clone())) {
+                            warn!(
+                                target: LOG_TARGET,
+                                "Could not restart transaction negotiation protocols: {}", e
```
This message doesn't seem related to the error.

SWvheerden committed:
Add tor and node identity to  sql lite storage

Add tor_id creation and storage to libwallet
update conversion error to not suppress error
